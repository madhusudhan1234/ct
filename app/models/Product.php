<?php

namespace App\Models;

class Product
{
    public static function getProducts(){
        return [
            [
                "name" => "Tshirt",
                "price" => "100",
                "currency" => "USD",
            ],
            [
                "name" => "Monitor",
                "price" => "100",
                "currency" => "USD",
            ]
        ];
    }
}