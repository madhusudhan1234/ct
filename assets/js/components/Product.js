import React, { Component } from 'react';

export default  class Product extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isShowForm: false,
      selectedQuantity: 1,
      selectedPayMethod: 'cash',
      totalPrice: null,
    }

    this.handlePayMethodChange = this.handlePayMethodChange.bind(this);
    this.showForm = this.showForm.bind(this);
    this.handleQuantityChange = this.handleQuantityChange.bind(this);
    this.calculatePrice = this.calculatePrice.bind(this);
  }

  handlePayMethodChange(e) {
    this.setState({ selectedPayMethod: e.target.value });
  }

  showForm() {
    this.setState({ isShowForm: true });
  };

  handleQuantityChange(e) {
    this.setState({ selectedQuantity: e.target.value }, () => this.calculatePrice());
  }

  calculatePrice() {
    this.setState({ totalPrice: this.props.product.price * this.state.selectedQuantity });
  }

  render(){
    return (
      <div className="card" key={this.props.product.id}>
        <div className="card-body">
          <h5 className="card-title">{this.props.product.name}</h5>
          <p className="card-text">{this.props.product.currency} {this.props.product.price}</p>
          {!this.state.isShowForm && <button className="btn btn-primary" onClick={this.showForm}>Process to Buy</button>}
          {this.state.isShowForm &&
            <form>
              <div className="form-group">
                <label htmlFor="payment">How do you pay?</label>
                <select className="form-control" id="payment" onChange={this.handlePayMethodChange}>
                  {this.props.paymentMethods.map(paymentMethod =>
                    <option key={paymentMethod.id} value={paymentMethod.type}>{paymentMethod.name}</option>
                  )}
                </select>
              </div>
              <div className="form-group">
                <label htmlFor="quantity">Quantity</label>
                <input className="form-control" type="number" name="quantity" defaultValue="1" min="1" max="100" onChange={this.handleQuantityChange} />
              </div>
              <button
                className="btn btn-primary"
                onClick={(e) => this.props.order(e, this.state.selectedPayMethod, this.state.selectedQuantity, this.props.product.price, this.props.product.name)}>
                Buy
              </button>
            </form>
          }
          {this.state.totalPrice && <span>Price: <strong>{this.props.product.currency} {this.state.totalPrice}</strong></span>}
        </div>
      </div>
    );
  }

  componentDidMount() {
    this.calculatePrice();
  }
}