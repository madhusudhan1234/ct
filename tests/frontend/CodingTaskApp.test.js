import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../../assets/js/CodingTaskApp';
import Product from '../../assets/js/components/Product';

global.fetch = require('jest-fetch-mock');
Enzyme.configure({ adapter: new Adapter() });

describe('CodingTaskApp', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Product Component', () => {
  const props = {
    product: {
      "id": "1",
      "name": "Tshirt",
      "price": "50",
      "currency": "USD"
    },
    paymentMethods: [
      {
        "id": "1",
        "type": "cash",
        "name": "Cash on Delivery"
      },
      {
        "id": "2",
        "type": "stripe",
        "name": "Stripe Payment Gateway"
      }
    ],
    order: jest.fn(),
  }

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Product { ...props } />, div);
  });

  test('Product has a valid snapshot', () => {
    const component = renderer.create(
        <Product { ...props } />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});