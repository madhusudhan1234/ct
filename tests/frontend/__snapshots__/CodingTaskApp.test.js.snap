// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodingTaskApp has a valid snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Available Products!
  </h1>
  
</div>
`;

exports[`Product Component Product has a valid snapshot 1`] = `
<div
  className="card"
>
  <div
    className="card-body"
  >
    <h5
      className="card-title"
    >
      Tshirt
    </h5>
    <p
      className="card-text"
    >
      USD
       
      50
    </p>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Process to Buy
    </button>
    <span>
      Price: 
      <strong>
        USD
         
        50
      </strong>
    </span>
  </div>
</div>
`;
